+ ECE + discontinuous regression

tennisonliu opened the pull request:
* + expected calibration error calculations
* - remove saved models
* + regression with artificial gap
* clean up tracking
* update README with extensions

Commit:
clean tracking

chelsealuisa commented on README.md:
```diff
@@ -7,6 +7,11 @@ Paper: Blundell, C., Cornebise, J., Kavukcuoglu, K. and Wierstra, D., 2015, June
         In International Conference on Machine Learning (pp. 1613-1622). PMLR.
        (http://proceedings.mlr.press/v37/blundell15.html)
 
+## Extension
+Additional approximate inference methods are implemented including:
+* *Bayes by Backprop - Local Reparameterisation Trick* (https://arxiv.org/pdf/1506.02557.pdf): Samples pre-activations instead of weights for lower variance, faster computation and convergence.
+* *Monte Carlo Dropout* (https://arxiv.org/pdf/1506.02142.pdf): Dropout during test-time to generate uncertainty measures, p=0.5.
+* *Functional Variational Inference* (https://arxiv.org/pdf/1903.05779.pdf): Optimises ELBO defined on stochastic processes, *i.e.* distribution over functions.
```
Nice!

chelsealuisa commented on README.md:
```diff
@@ -23,5 +28,10 @@ At run-time, `main` reads from a model configuration set in `config.py`. The con
+## Analysis:
+The scripts `weight_pruning.py` and `compute_ece.py` perform post-hoc analysis using saved models.
+* `weight_pruning` 1) plots the distribution of weights, 2) computes SNR of BNNs, 3) evaluates performance on pruned weights.
+* `compute_ece` 1) computes the ECE of trained model, 2) plots reliability diagram.
```
Stupid question - what does ECE stand for?

tennisonliu replied:
done

chelsealuisa commented on classification/class_task.py:
```diff
@@ -165,7 +165,7 @@ def evaluate(self, test_loader):
                 X, y = data
                 X, y = X.to(DEVICE), y.to(DEVICE)
-                preds = self.predict(X)
+                preds, _ = self.sample_predict(X)
```
Possibly way too nit-picky, but should we instead change all of the sample_predict methods to just predict()? sample_predict doesn't really make sense for MLP

tennisonliu replied:
done.

chelsealuisa commented on config.py:
```diff
@@ -13,12 +13,12 @@ class RegConfig:
-    test_samples = 100
+    test_samples = 10
     num_test_points = 400
     mode = 'regression'
     mixture_prior = False
-    hidden_units = 400
-    noise_tolerance = .1
+    hidden_units = 600              # CHANGED TO 100
```
Could we leave the config settings as they were to reflect the values that were used to create the final saved models/results that will be included in the report?

tennisonliu replied:
done.

chelsealuisa commented on compute_ece.py:
```diff
@@ -0,0 +1,136 @@
+    config = ClassConfig
+    test_ds = create_data_class(train=False, batch_size=config.batch_size, shuffle=False)
+
+    params = {
+        'lr': config.lr,
+        'hidden_units': config.hidden_units,
+        'mode': config.mode,
+        'batch_size': config.batch_size,
+        'epochs': config.epochs,
+        'x_shape': config.x_shape,
+        'classes': config.classes,
+        'num_batches': 0,
+        'train_samples': config.train_samples,
+        'test_samples': 5,
+        'mu_init': config.mu_init,
+        'rho_init': config.rho_init,
+        'prior_init': config.prior_init,
+        'mixture_prior':config.mixture_prior,
+        'save_dir': config.save_dir,
+    }
+
+    models = {
+        'BNN': BNN_Classification('bnn_classification', {**params, 'local_reparam': False, 'dropout': False}),
+        'BNN-LR': BNN_Classification('bnn_classification_lr', {**params, 'local_reparam': True, 'dropout': False}),
+        'MLP': MLP_Classification('mlp_classification', {**params, 'dropout': False}),
+        # 'dropout_class': MLP_Classification('dropout_classification', {**params, 'dropout': True}),
+        'MC-Dropout': MCDropout_Classification('mcdropout_classification', {**params, 'dropout': True}),
+        }
+
+    for _, model in models.items():
+        model.net.load_state_dict(torch.load(model.save_model_path))
```
Could this be replaced with the functions in load_model_utils instead?

tennisonliu replied:
That's what I thought to do, but we need functions e.g. predict() and enable_dropout() that are only available in the class_task objects and not the models.

chelsealuisa commented on load_model_utils.py:
```diff
@@ -1,78 +1,9 @@
-        return x
-
-    def log_prior(self):
-        return self.l1.log_prior + self.l2.log_prior + self.l3.log_prior
-    
-    def log_variational_posterior(self):
-        return self.l1.log_variational_posterior + self.l2.log_variational_posterior + self.l3.log_variational_posterior
-
-    def kl_cost(self):
-        return self.l1.kl_cost + self.l2.kl_cost + self.l3.kl_cost
-
-    def get_nll(self, outputs, target, sigma=1.):
-        if self.mode == 'regression':
-            nll = -torch.distributions.Normal(outputs, sigma).log_prob(target).sum()
-        elif self.mode == 'classification':
-            nll = nn.CrossEntropyLoss(reduction='sum')(outputs, target)
-        else:
-            raise Exception("Training mode must be either 'regression' or 'classification'")
-        return nll
-
-    def sample_elbo(self, input, target, beta, samples, sigma=1.):
-        ''' Sample ELBO for BNN w/o Local Reparameterisation '''
-        assert self.local_reparam==False, 'sample_elbo() method returns loss for BNNs without local reparameterisation, alternatively use sample_elbo_lr()'
-        log_priors = torch.zeros(samples).to(DEVICE)
-        log_variational_posteriors = torch.zeros(samples).to(DEVICE)
-        negative_log_likelihood = torch.zeros(1).to(DEVICE)
-
-        for i in range(samples):
-            output = self.forward(input, sample=True)
-            log_priors[i] = self.log_prior()
-            log_variational_posteriors[i] = self.log_variational_posterior()
-            negative_log_likelihood += self.get_nll(output, target, sigma)
-
-        log_prior = beta*log_priors.mean()
-        log_variational_posterior = beta*log_variational_posteriors.mean()
-        negative_log_likelihood = negative_log_likelihood / samples
-        loss = log_variational_posterior - log_prior + negative_log_likelihood
-        return loss, log_priors.mean(), log_variational_posteriors.mean(), negative_log_likelihood
-
-def load_bnn_class_model(saved_model):
```
thx for removing. This was a temporary addition to be able to work with the older saved models that had different model params, I should've removed it already.

chelsealuisa commented on main.py:
```diff
@@ -13,7 +13,7 @@
     ''' Regression Task Trainer '''
     config = RegConfig
-    X, Y = create_data_reg(train_size=config.train_size)
+    X, Y = create_data_reg(train_size=config.train_size, gap=True)
```
Shouldn't this be set to False as a default?

tennisonliu replied:
moved this into configuration file.

Commit:
+ final plots + poster + move helper functions into /utils

tennisonliu committed:
Update README.md

tennisonliu committed:
Create README.md

tennisonliu committed:
+ relative path to poster

Commit:
remove raw downloads

Commit:
update relative paths

tennisonliu commented:
@chelsealuisa made the following changes:
1) uploaded (final) plots in graphs/final_graphs
2) removed saved models from repo
3) updated readme with poster link
4) moved util functions into /utils folder (along with necessary refactoring to change relative paths)

chelsealuisa commented:
Looks great!
